### PR TITLE
Show current project context for help command

### LINKED ISSUE #557: Show current project context for help command
When running `shopify help`, it would be good to have a reminder of what context you're in so it's clearer why the list of commands changes. So if you're inside a project, it should say something like:

```console
$ shopify help
CLI to help build Shopify apps faster.

Use shopify help <command> to display detailed information about a specific command.

Project: my_project (Node app)
Available commands for Node app projects:
```

/cc @gfscott @iainmcampbell 

We may also went to extend this contextual reminder to other commands besides help if their behavior differs across project types or depending on whether you're in a project or not.

### PR DESCRIPTION
**Pull request by andyw8**
### WHY are these changes introduced?

Fixes #557

### WHAT is this pull request doing?

We're partitioning the command list into two categories, core and local. Core commands are available everywhere, local commands are only available when inside a supported app's directory, i.e. Node or Rails currerently.

Before:

Screen Shot 2020-05-04 at 4 01 04 PM

After:

Screen Shot 2020-05-06 at 10 18 10 AM

- [x] Fix failing tests
- [x] Add 'after' screenshot
- [x] Add project indicator, e.g. `Project: my_project (Node app)`
- [x] Verify with a sample Rails app
- [x] Verify with a sample Node app
- [x] Verify with a sample unsupported app
- [x] Update screenshot to show latest output (including project path)
- [x] Update to use core check from #568
- [x] Update for changes in #574
- [ ] Squash before merging to master

### CODE REVIEW

File: lib/shopify-cli/commands/help.rb [hunk 1]
```diff
@@ -21,27 +21,45 @@ def call(args, _name)
         end
 
         preamble = <<~MESSAGE
-          CLI to help build Shopify apps faster.
```
Comment (andyw8) [inline lib/shopify-cli/commands/help.rb:24]: Remove as per @gfscott's suggestion: https://github.com/Shopify/shopify-app-cli/issues/557#issuecomment-623485924

File: lib/shopify-cli/commands/help.rb [hunk 2]
```diff
@@ -21,27 +21,45 @@ def call(args, _name)
...
       private
 
+      def project_name
+        ProjectType.load_type(Project.current_project_type).project_name
+      end
+
+      def core_commands
+        resolved_commands
+          .select { |_name, c| !c.hidden }
+          .select { |_name, c| c.to_s.include?('ShopifyCli::Commands') }
```
Comment (andyw8) [inline lib/shopify-cli/commands/help.rb:55]: I'm still thinking about a better way to distinguish core vs local commands.

Comment (mkevinosullivan) [reply lib/shopify-cli/commands/help.rb:55]: I'll be committing a `ShopifyCli::Commands.core_command?` method shortly as part of another fix, I'll add you as a reviewer

File: lib/shopify-cli/commands/help.rb [hunk 3]
```diff
@@ -21,27 +21,45 @@ def call(args, _name)
...
-        visible_commands = ShopifyCli::Commands::Registry
-          .resolved_commands
-          .select { |_name, c| !c.hidden }
-          .sort
+        core_commands.each do |name, klass|
+          next if name == 'help'
+          @ctx.puts("{{command:#{name}}}: #{klass.help}\n")
+        end
+
+        return unless Project.current_project_type && ProjectType.load_type(Project.current_project_type)
```
Comment (andyw8) [inline lib/shopify-cli/commands/help.rb:36]: This can perhaps be simplified.

File: lib/shopify-cli/commands/help.rb (hunk 2 above)
Comment (andyw8) [reply lib/shopify-cli/commands/help.rb:55]: Also, I'm thinking this should probably be extracted out of the `Command` class and into something like a registry, which could have query methods for the core and local commands. But that could be done in a follow-up PR.

File: test/shopify-cli/commands/help_test.rb [hunk 4]
```diff
@@ -25,10 +41,59 @@ def test_default_behavior_lists_tasks
...
+          run_cmd('help')
+        end
+        output = io.join
+
+        refute_match(/Available commands.*Rails/, output)
+        refute_match(/fake_rails/, output)
+      end
+
+      def test_shows_current_project_path_and_type
+        Dir.stubs(:pwd).returns("/Users/john/my_app")
```
Comment (andyw8) [inline test/shopify-cli/commands/help_test.rb:85]: I would like to have a better way to do this without having to stub `Dir`.

File: test/shopify-cli/commands/help_test.rb [hunk 5]
```diff
@@ -25,10 +41,59 @@ def test_default_behavior_lists_tasks
...
+        end
+        output = io.join
+
+        assert_match('Available core commands:', output)
+        assert_match(/Usage: .*shopify/, output)
+      end
+
+      def test_local_commands_available_within_a_project
+        Project.stubs(:current_project_type).returns('rails')
+        Registry.add(->() { Rails::Commands::Fake }, 'fake_rails')
```
Comment (andyw8) [inline test/shopify-cli/commands/help_test.rb:60]: @mkevinosullivan I've updated this to make use of the new behaviour you've added, but I realise there's maybe some inconsistency.

To add a new core command, we call `Commands.register`, which then calls `Registry.add`.
To add a new local command, we call `Registry.add` directly, e.g. as in `ProjectType.register_command`.

I wonder if would make sense to split `Commands.register` into `Commands.register_core` and `Commands.register_local` to make the intentional clearer?

Comment (mkevinosullivan) [reply test/shopify-cli/commands/help_test.rb:60]: It's a fair point.  There's a bit of abstraction missing here.  The `register_command` within `ShopifyCli::ProjectType` knows that the Registry that's within `ShopifyCli::Commands` is of type `CLI::Kit::Registry` and accepts the `add` method/message.  It really should be using `ShopifyCli::Commands.register()` method, rather than calling `Registry.add()` directly.

I'll create a new PR for review to:
- add an `is_core` parameter to the register() method, default to `false`
- update the calls to `register` within `ShopifyCli::Commands` to include `true` at the end
- change the `register_command` method in `ProjectType` to use the Commands.register() method instead (thereby abstracting the underlying Registry).

Makes sense?

Comment (andyw8) [reply test/shopify-cli/commands/help_test.rb:60]: Sounds good, but I think I'd lean towards a separate `register_core` method, following Fowler's advice for avoiding flag arguments:
https://learning.oreilly.com/library/view/refactoring-improving-the/9780134757681/ch11.xhtml#ch11lev1sec3

File: test/shopify-cli/commands/help_test.rb [hunk 6]
```diff
@@ -25,10 +41,47 @@ def test_default_behavior_lists_tasks
...
+
+        io = capture_io do
+          run_cmd('help')
+        end
+        output = io.join
+
+        assert_match(/Available commands for Ruby on Rails App projects.*fake_rails/m, output)
```
Comment (andyw8) [inline test/shopify-cli/commands/help_test.rb:57]: I'm using `/m` to match across multiple lines, but this isn't ideal, I'd rather test against the structure than the presentation. Hopefully we can find a way to untangle these later.

File: lib/shopify-cli/commands/help.rb [hunk 7]
```diff
@@ -21,27 +21,46 @@ def call(args, _name)
...
+        core_commands.each do |name, klass|
+          next if name == 'help'
+          @ctx.puts("{{command:#{name}}}: #{klass.help}\n")
+        end
+
+        return unless inside_supported_project?
 
-        visible_commands.each do |name, klass|
+        @ctx.puts("{{bold:Project: #{File.basename(Dir.pwd)} (#{project_name})}}")
+        @ctx.puts("{{bold:Available commands for #{project_name} projects:}}\n\n")
```
Comment (mkevinosullivan) [inline lib/shopify-cli/commands/help.rb:39]: Minor quibble - `project_name` might be interpreted to allude to the name of the project (as provided by the developer) whose directory you're actually in, vs. the name of the project type of the project whose directory you're in!  So, maybe `project_type_name`?

Comment (andyw8) [reply lib/shopify-cli/commands/help.rb]: Yup, i agree. Fixed in 899da5e

Comment (mkevinosullivan) [reply lib/shopify-cli/commands/help.rb]: 👍

File: lib/shopify-cli/commands/help.rb [hunk 8]
```diff
@@ -21,27 +21,46 @@ def call(args, _name)
...
+          next if name == 'help'
+          @ctx.puts("{{command:#{name}}}: #{klass.help}\n")
+        end
+
+        return unless inside_supported_project?
 
-        visible_commands.each do |name, klass|
+        @ctx.puts("{{bold:Project: #{File.basename(Dir.pwd)} (#{project_type_name})}}")
+        @ctx.puts("{{bold:Available commands for #{project_type_name} projects:}}\n\n")
+
```
Comment (mkevinosullivan) [inline lib/shopify-cli/commands/help.rb:40]: `#{project_type_name}` is shown on two consecutive lines, which seems a little redundant.  Maybe something like:
```
@ctx.puts("{{bold:Available commands for #{project_type_name} project}} {{cyan:#{File.basename(Dir.pwd)}:}}\n\n")
```
or
```
@ctx.puts("{{bold:Available commands for}} {{cyan:#{File.basename(Dir.pwd)} (#{project_type_name}):\n\n")
```

Also (perhaps make an issue for this for another PR), it would be nicer to have something like `Project.project_name` (for example) instead of `File.basename(Dir.pwd)` to convey the project name.

Comment (andyw8) [reply lib/shopify-cli/commands/help.rb:40]: Any thoughts on this @attaboy?

> Also (perhaps make an issue for this for another PR), it would be nicer to have something like Project.project_name (for example) instead of File.basename(Dir.pwd) to convey the project name.

Yup, I agree, I'll handle this in a follow-up.

Comment (attaboy) [reply lib/shopify-cli/commands/help.rb:40]: It is a little redundant agree, but I think it's useful redundancy — we're giving the user some crucial information here, and helping reinforce some concepts that I don't think are terribly obvious — that there is a concept of a "project", and there are different types of projects, and that different types of projects have specific commands.

I considered `Available commands for this project type:` for the second line, but IMO it's just slightly less clear. (I don't feel that strongly about it though.)

Comment (mkevinosullivan) [reply lib/shopify-cli/commands/help.rb:40]: Fair enough - if, from a UX perspective, it's useful to keep it to convey a message/concept, I'm good.
